File: tools/caatuu-cz-ml/scripts/ml/compare_benchmarks.py
#!/usr/bin/env python3
"""Create a compact side-by-side benchmark report."""

from __future__ import annotations

import argparse
import json
from pathlib import Path


def read_json(path: Path):
    return json.loads(path.read_text(encoding="utf-8"))


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--base", required=True)
    parser.add_argument("--finetuned", required=True)
    parser.add_argument("--out", required=True)
    args = parser.parse_args()

    base = read_json(Path(args.base))
    tuned = read_json(Path(args.finetuned))
    tuned_by_id = {row["id"]: row for row in tuned["prompts"]}

    lines = [
        "# Qwen3-1.7B Czech LoRA Benchmark",
        "",
        f"- Base model: `{base['model_id']}`",
        f"- Adapter: `{tuned['adapter']}`",
        f"- Device: `{tuned['device']}`",
        f"- Base total generation time: `{base['total_seconds']}s`",
        f"- Fine-tuned total generation time: `{tuned['total_seconds']}s`",
        "",
        "## Side-by-side",
        "",
    ]

    for base_row in base["prompts"]:
        tuned_row = tuned_by_id[base_row["id"]]
        lines.extend(
            [
                f"### {base_row['id']}",
                "",
                f"Prompt: {base_row['user']}",
                "",
                "Base:",
                "",
                "```text",
                base_row["output"],
                "```",
                "",
                "Fine-tuned:",
                "",
                "```text",
                tuned_row["output"],
                "```",
                "",
            ]
        )

    lines.extend(
        [
            "## Notes",
            "",
            "- The adapter improved direct diacritic restoration in several cases.",
            "- The adapter is not production-ready: it still makes Czech lexical errors and collapsed the open shop-dialogue prompt.",
            "- This run is useful as proof that LoRA training works on the local GPU, not as the final Czech model.",
            "",
        ]
    )
    Path(args.out).write_text("\n".join(lines), encoding="utf-8")
    print(args.out)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/czech-ml/scripts/compare-benchmarks.mjs
#!/usr/bin/env node
import fs from "node:fs/promises";
import { readJson } from "./jsonl.mjs";

function arg(name, fallback = null) {
  const idx = process.argv.indexOf(name);
  return idx >= 0 ? process.argv[idx + 1] : fallback;
}

const basePath = arg("--base");
const tunedPath = arg("--finetuned");
const outPath = arg("--out");

if (!basePath || !tunedPath || !outPath) {
  console.error("Usage: node scripts/compare-benchmarks.mjs --base base.json --finetuned tuned.json --out report.md");
  process.exit(2);
}

const base = await readJson(basePath);
const tuned = await readJson(tunedPath);
const tunedById = new Map(tuned.prompts.map((row) => [row.id, row]));

const lines = [
  "# Qwen3-1.7B Czech LoRA Benchmark",
  "",
  `- Base model: \`${base.model_id}\``,
  `- Adapter: \`${tuned.adapter}\``,
  `- Device: \`${tuned.device}\``,
  `- Base total generation time: \`${base.total_seconds}s\``,
  `- Fine-tuned total generation time: \`${tuned.total_seconds}s\``,
  "",
  "## Side-by-side",
  ""
];

for (const baseRow of base.prompts) {
  const tunedRow = tunedById.get(baseRow.id);
  lines.push(
    `### ${baseRow.id}`,
    "",
    `Prompt: ${baseRow.user}`,
    "",
    "Base:",
    "",
    "```text",
    baseRow.output || "",
    "```",
    "",
    "Fine-tuned:",
    "",
    "```text",
    tunedRow?.output || "",
    "```",
    ""
  );
}

await fs.mkdir(outPath.replace(/[\\/][^\\/]*$/, ""), { recursive: true });
await fs.writeFile(outPath, lines.join("\n"), "utf8");
console.log(outPath);
